### PR TITLE
Add +tools/cquery layer

### PR DESCRIPTION
**Pull request by MaskRay**
tl;dr https://github.com/cquery-project/cquery/wiki/Emacs
https://www.reddit.com/r/emacs/comments/806mgw/recent_improvement_of_cquery_and_emacscquery/

[cquery](https://github.com/cquery-project/cquery) is a low-latency language server supporting multi-million line C++ code-bases, powered by libclang.

Since [current c-c++ layer needs cleanup](https://github.com/syl20bnr/spacemacs/issues/10143) and I am not an experienced elisp or spacemacs user, I decide to put it into `+tools/` instead of `+lang/c-c++/`. Another reason doing this is that cquery also supports Objective-C as libclang supports it, though probably with many issues because of the lack of Objective-C users 😜 Anyway, I hope more experienced user can move it into `+lang/c-c++` or elsewhere that is appropriate.

+tools/cquery depends on +tools/lsp, which is added in another PR https://github.com/syl20bnr/spacemacs/pull/10211

### CODE REVIEW

**Comment by Jackson-soft** [comment]
good!!

**Comment by MaskRay** [comment]
Some tricky parts. cquery saves a copy of indexed source files (C/C++/Objective-C) in `.cquery_cached_index` (relative to the project root by default) and these files may be listed by projectile.

Adding the cache directory to `projectile-globally-ignored-directories` works but if the user customizes `cquery-cache-dir`, the logic blacklist directory should be changed accordingly.
```elisp
    (require 'projectile)
    (add-to-list 'projectile-globally-ignored-directories ".cquery_cached_index")
```

cquery users usually do not use ycmd for completion while ycmd is enabled by default in `+lang/c-c++/packages.el`. I personally use `   dotspacemacs-excluded-packages '(cmake-ide info+ racer ycmd)`

**Comment by sdwolfz** [comment]
How does cquery behave when the `.cquery_cached_index` directory is located in the project directory and you execute `git clean -fdx`? Does it know to rebuild it or does it glitch out / crash?

**Comment by MaskRay** [comment]
> How does cquery behave when the .cquery_cached_index directory is located in the project directory and you execute git clean -fdx? Does it know to rebuild it or does it glitch out / crash?

@jacobdufault to answer this question.

**Comment by MaskRay** [comment]
I hope someone could pick this up and make +lang/c-c++ layer accept cquery as a backend.

**Comment by iromise** [comment]
Hope it can be integrated in spacemacs soon.

**Comment by cormacc** [comment]
Reworked some of MaskRay's config into a a new PR for the lsp layer, here:
https://github.com/syl20bnr/spacemacs/pull/10486

And an attempt at a cquery layer here:
https://github.com/cormacc/spacemacs/commit/b8ae1bc3f8e19db53f4badb8c11110d746128494

I'll raise a new PR for the cquery layer if the lsp-layer one gets merged (it builds on it).
